prefect_dbt/cloud/jobs.py

```diff
@@ -318,9 +326,182 @@ def my_flow():
             f"Triggered job run with ID {run_id} was cancelled."
         )
     elif final_run_status == DbtCloudJobRunStatus.FAILED:
-        raise DbtCloudJobRunFailed(f"Triggered job run with ID: {run_id} failed.")
+        while retry_filtered_models_attempts > 0:
+            logger.info(
+                f"Retrying job run with ID: {run_id} "
+                f"{retry_filtered_models_attempts} more times"
+            )
+            try:
+                retry_filtered_models_attempts -= 1
+                run_data = await (
+                    retry_dbt_cloud_job_run_subset_and_wait_for_completion(
+                        dbt_cloud_credentials=dbt_cloud_credentials,
+                        run_id=run_id,
+                        trigger_job_run_options=trigger_job_run_options,
+                        max_wait_seconds=max_wait_seconds,
+                        poll_frequency_seconds=poll_frequency_seconds,
+                    )
+                )
+                return run_data
+            except Exception:
+                pass
+        else:
+            raise DbtCloudJobRunFailed(f"Triggered job run with ID: {run_id} failed.")
     else:
         raise RuntimeError(
             f"Triggered job run with ID: {run_id} ended with unexpected"
             f"status {final_run_status.value}."
         )
+
+
+async def _build_trigger_job_run_options(
+    dbt_cloud_credentials: DbtCloudCredentials,
+    trigger_job_run_options: TriggerJobRunOptions,
+    run_id: str,
+    run_info: Dict[str, Any],
+):
+    """
+    Compiles a list of steps (commands) to retry, then either build trigger job
+    run options from scratch if it does not exist, else overrides the existing.
+    """
+    steps_override = []
+    for run_step in run_info["run_steps"]:
+        status = run_step["status_humanized"].lower()
+        # Skipping cloning, profile setup, and dbt deps - always the first three
+        # steps in any run, and note, index starts at 1 instead of 0
+        if run_step["index"] <= 3 or status == "success":
+            continue
+        # get dbt build from "Invoke dbt with `dbt build`"
+        command = run_step["name"].partition("`")[2].partition("`")[0]
+        sub_command = shlex.split(command)[1]
+
+        is_run_command = command in ("build", "run", "test", "seed", "snapshot")
```

Comment:
A job within dbt Cloud is essentially a collection of dbt commands.  The job will iterate over those commands in succession.  However, in the event a step fails, any subsequent steps will be skipped (unless we've checked Generate Docs, which will always run a `dbt docs generate` command regardless of status from a previous step).

So, the first if statement is addressing whether or not the step is one where we need to find individual nodes that failed/skipped.  Otherwise, we just need to rerun that step entirely.  And, we only need to find the individual nodes for those run commands (other commands like `run-operation` are one's we need to rerun entirely).